### PR TITLE
variables: Enable dynamic evaluation with customizable function

### PR DESCRIPTION
**Pull request by theHamsta**
New default for evaluation expression is visual selection or `cword`

This enable inspecting of member functions without explicit `this`

Screenshot_20210202_203301

Or evaluation from visual selection

image
when selecting `this->m_member`
image

The functions for visual selection come from nvim-treesitter. I wrote the one that gives you the text from the range.

You can replace the function for the current expression to evaluate if you want to.

### CODE REVIEW

File: lua/dap/ui/variables.lua [hunk 1]
```diff
@@ -230,20 +233,32 @@ function M.hover()
...
         break
       end
     end
+
     if variable then
       floating_win, floating_buf = popup()
       create_variable_buffer(floating_buf, floating_win, session, {variable})
     else
-      print('"'..cword..'" not found!')
+      session:evaluate(expression, function(err, resp)
```
Comment (mfussenegger) [inline lua/dap/ui/variables.lua:250]: oh nice :+1:

**Review by mfussenegger — APPROVED** [review]
Nice work :+1:

**Review by mfussenegger — COMMENTED** [review]
One thing that may be a bit strange is that selection marks are not really cleared, so you may get the visual selection behavior even if you want to have the word under the cursor.

Might make sense to check the mode or otherwise add a dedicated `visual_hover` variant?

**Comment by theHamsta** [comment]
I had a solution with `nvim_get_mode` but this somehow does not really work in Lua mode.

**Comment by theHamsta** [comment]
I'll add visual_hover...

**Pull request merged**